data/impl/StoreValidator.js

```diff
@@ -0,0 +1,187 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2021 Extremely Heavy Industries Inc.
+ */
+
+import {HoistBase, managed} from '@xh/hoist/core';
+import {action, computed, makeObservable, observable} from '@xh/hoist/mobx';
+import {PendingTaskModel} from '@xh/hoist/utils/async';
+import {cloneDeep, compact, flatten, isEmpty, map} from 'lodash';
+
+import {ValidationState} from '../validation/ValidationState';
+
+/**
+ * Computes and stores validation state for a Store and its modified Records.
+ *
+ * @private
+ */
+export class StoreValidator extends HoistBase {
+
+    /** @member {Store} */
+    store;
+
+    @managed _validationTask = new PendingTaskModel({mode: 'all'});
+
+    // Stores the result of evaluating each record. Errors are store in a deep map structure
+    // to facilitate access at the record, field and rule levels:
+    // e.g. _errors[record.id][field.name][rule] = errors[]
+    @observable.ref _errors = {};
```

Comment:
We could consider adding a RecordErrors impl class that might be the thing that is stored for each record.